### PR TITLE
Add specific docstrings to Int and UInt

### PR DESCRIPTION
**Pull request by jakobnissen**
Anecdotally, it is not always clear for beginners that `Int` and `Int64` is the same type, and this leads to some confusion for newcomers. This is described in the manual, but it would be nice to also have it in the docstring.

Furthermore, mention that `Int` is used as the default integer type in Julia. Where some languages use `UInt` for things such as array sizes, Julia nearly universally uses `Int` as the integer type.

### CODE REVIEW

File: base/docs/basedocs.jl [hunk 1]
```diff
@@ -1999,6 +1999,39 @@ for bit in (8, 16, 32, 64, 128)
     end
 end
 
+"""
+	$(Int) <: Signed
+
+`Int` is aliased to `Int64` on 64-bit systems and `Int32` on 32-bit systems.
```
Comment (giordano) [inline base/docs/basedocs.jl:2005]: Can you actually access this docstring and `Int64`'s independently?

**Comment by mcabbott** [comment]
See also https://github.com/JuliaLang/julia/pull/45221 which mentions that there is this default for `Int` (although not at present for `UInt`)

File: base/docs/basedocs.jl (hunk 1 above)
Comment (jakobnissen) [reply base/docs/basedocs.jl:2005]: Not that I know of. This shows up the docstring for Int64.

Comment (giordano) [reply base/docs/basedocs.jl:2005]: So you get both of them or this shadows the other one?

Comment (jakobnissen) [reply base/docs/basedocs.jl:2005]: It shadows.

**Comment by jakobnissen** [comment]
CI failure seems unrelated.

File: base/docs/basedocs.jl [hunk 2]
```diff
@@ -1999,6 +1999,39 @@ for bit in (8, 16, 32, 64, 128)
     end
 end
 
+"""
+	$(Int) <: Signed
```
Comment (aviatesk) [inline base/docs/basedocs.jl:2003]:
```suggestion
	Int <: Signed
```
Isn't more reasonable if we hardcode this?

**Review by aviatesk — APPROVED** [review]

**Comment by mcabbott** [comment]
If this is approved, can you clarify what you envisage happening with the earlier https://github.com/JuliaLang/julia/pull/45221? That also aims to explain the Int default, but does not make a separate `Int` docstring, as I thought that would overwrite the one for `Int64`.

**Comment by aviatesk** [comment]
It seems like we can access to both documentations even if we have a separate documentation for `Int`:
```julia
julia> @eval begin
           """
               Parent
       
           Documentation for `Parent`
           """
           struct Parent end
       
           """
               Aliased
       
           Documentation for `Aliased`
           """
           const Aliased = Parent
       end
Aliased

help?> Parent
search: Parent parent parentmodule parentindices

  Parent


  Documentation for Parent

help?> Aliased
search: Aliased

  Aliased


  Documentation for Aliased
```

**Comment by jakobnissen** [comment]
This is from a fresh build off this branch. The docstring only appears once.

image

**Comment by jakobnissen** [comment]
Oh sorry, you are right! I did not understand Shuhei's example. Yeah, I agree, this is not what I want here. There should only be one docstring, `Int64`, and accessing `Int` should display that.
I'll fix it shortly.

**Comment by mcabbott** [comment]
> There should only be one docstring, Int64, and accessing Int should display that.

This is exactly what https://github.com/JuliaLang/julia/pull/45221 does, it inserts a block about `Int` into either the `Int64` or the `Int32` docstring. Maybe it should be longer, give examples... can I suggest adding to that?

**Comment by ViralBShah** [comment]
Is this good to merge? I am rebasing it to master to be able to run CI (which was quite broken the last time it ran here).

**Comment by jakobnissen** [comment]
@ViralBShah I see there are some CI failures with building the docs. I'll iron them out, then ping you when tests pass and it can be merged

**Comment by jakobnissen** [comment]
@ViralBShah After having skimmed them, I think the numerous CI failures are all unrelated to this change. This is ready to merge.

**Comment by ViralBShah** [comment]
@giordano @mcabbott Since you guys have both chimed in - do you have any thoughts on merging?

File: base/docs/basedocs.jl [hunk 3]
```diff
@@ -2028,6 +2030,62 @@ for bit in (8, 16, 32, 64, 128)
...
+# Examples
+```jldoctest
+julia> 1 isa Int
+true
+
+julia> typeof(length(Dict())) === Int
+true
+
+julia> Threads.nthreads() isa Int
+true
```
Comment (halleysfifthinc) [inline base/docs/basedocs.jl:2075]: These examples are a bit obscure/advanced for a docstring about such a basic, fundamental type (especially the `nthreads()` example. Instead, maybe just:
```suggestion
julia> typeof(length([1,2])) === Int
true
```

Comment (ViralBShah) [reply base/docs/basedocs.jl:2075]: Good idea not to introduce multi-threading here.

**Comment by mcabbott** [comment]
It still seems confusing to me to have two docstrings for `Int` and `Int64` when these are `===`. I have not built this branch, and am not quite sure whether the claim above is that the two may both be accessed (as in [the example code](https://github.com/JuliaLang/julia/pull/45415#issuecomment-1143084884)) which is confusing, or whether one is completely overwritten. But if completely overwritten, then it still seems a confusing way to write the source code, like defining a function twice & expecting the reader to know which will win. I thought [this comment](https://github.com/JuliaLang/julia/pull/45415#issuecomment-1143765487) indicated an intention to remove one of them, but as far as I can see the code at present still contains two. 

I also still wonder what the plan is with the earlier PR #45221, which also expands the docstring for `Int64` to mention the alias `Int`. It alters the loop which defines this and `Int32` to add the mention to one of the two. My vote would be to just add whatever details are missing from that, there or in a follow-up PR.

**Comment by fingolfin** [comment]
Since PR #45221 has been merged, there are docstrings for `UInt` and `Int`. So I think this PR can be closed.

(If you'd like to tweak the new docstrings, I'd recommend opening a new PR to make sure it gets proper attention)

**Pull request closed without merging**